### ISSUE TITLE
Rule to forbid calling components as functions?

### ISSUE THREAD

**Issue by FezVrasta** [jsx-eslint/eslint-plugin-react#3208]
Sometimes inexperienced developers define functions that return jsx that would better work as components.

example:

```
const App = () => {
  function renderIcon() {
    return <i />;
  }
  
  return (
    <div>{renderIcon()}</div>
  );
}
```

I’d like a rule to forbid this behavior and instead suggest to convert it to a proper competent called with jsx.

this allows React to optimize its re-renders.

**Comment by AriPerkkio** [comment]
At least in the example `App` component the internal `renderIcon` function should not be converted to React component. That would violate [`react/no-unstable-nested-components`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unstable-nested-components.md) and produce even worse performance. 

But sure, moving it outside `App` component and declaring as proper React component would likely be better.

**Comment by FezVrasta** [comment]
Hi, yeah of course it would need to be moved to the root scope.

Is there any rule that allows to enforce what I'm describing already?

**Comment by ljharb** [comment]
In this case it would work better as just inline jsx, and there’s no real harm from putting that in the render function. Do you have some examples that are actually problematic?

**Comment by FezVrasta** [comment]
That would work too @ljharb , it really depends case by case, sometimes a component makes more sense for readability or to keep the input/output clearer.

I think the rule should just forbid the usage I showed above.

**Comment by FezVrasta** [comment]
The issue here is React specific because the problem is the usage of a function instead of a component, making it impossible for the reconciler to apply its optimizations.

**Comment by ljharb** [comment]
Right, but this case shouldn't BE a component - and there's no actual advantage to making it one.

**Comment by AriPerkkio** [comment]
I think this rule would still be useful. I've also seen these cases where developers create additional functions just to reduce the length of return statement's JSX block. Inline JSX or converting the function into React component would be better. 

For implementation I think this rule should start from `CallExpression`s inside JSX. Then look for scope to see if this function was declared within the scope.
This would prevent falsely flagging renderProps pattern.  

```jsx
01 | function renderFooter() {
02 |     return <div />;
03 | }
04 | 
05 | function MyComponent({ renderHeader }) {
06 |     return (
07 |         <>
08 |             {renderHeader()} // OK
09 |             <div>My Component</div>
10 |             {renderFooter()} // Report as error
11 |         </>
12 |     )
13 | }
```

Line 8 should not be reported as error since `renderHeader` is not declared in the scope. Line 10 should be since in its scope there is `FunctionDeclaration` for `renderFooter` present. I don't think this rule should care whether function is declared inside component body or not. 

So would this rule be "In JSX, prefer React components or elements instead of plain functions returning JSX"?

**Comment by ljharb** [comment]
I'm still not clear on the downsides here.

In that example, repeating `<div />` inline, or storing it in a variable and repeating the variable, would be identical. Making a separate Footer component and repeating it twice would be identical - except that it'd have an extra level on the component stack (which, to be fair, is negligible, but if anything it's a vote against it).

**Comment by andlbrei** [comment]
Kent C. Dodds has a blog post about this.
Maybe it contains some information relevant to this issue: https://kentcdodds.com/blog/dont-call-a-react-function-component

**Comment by ljharb** [comment]
That blog post is correctly describing why you shouldn't call a *react component* directly as a function.

It is not, however, talking about calling a _non-component function that returns JSX_ - typically called a "render function", or used as a prop, a "render prop".

I don't think it makes sense to have a rule forbidding this. If someone wants to share code examples that are legitimately problematic - as opposed to just "maybe that's not the best way to organize that code" - I'm sure I'd be convinced to reopen this.

**Comment by dacunaDesp** [comment]
I would like to re-open this issue. I found many codebases written with that rule activeted that is writen in that way.

```jsx
function Uploader ({ readOnly }) {
  function readOnlyMessage() {
    // checking some state and render jsx or null
  }

  function readOnlyFileList() {
    // render jsx
  }

  function renderPreloadedItems() {
   return scopedState.map(something => {
      return <Wrapper {...aLotOfPropsConfig}>{readOnlyFileList()}</Wrapper>
   })
  }
  
  function renderUploader() {
   return (
     <UiComponent {...propsAssignedBasedOfState}>
     {renderPreloadedItems()}
     {scopedState.map(moreJSXHere)}
     </UiComponent>
  )
  }

  return readOnly ? (
      <>
        {readOnlyMessage()}
        {renderPreloadedItems()}
      </>
    ) : (
      renderUploader()
    );
}
```

If this functions are written with title case like `RenderUploader` it will trigger a error, but for now it will not and it allowed to the unexperience dev to bypass the rule. I try to disable the allowedProp to see if in that case it will be throwing some error but still taking this as valid code. I also create a unit test in this repo and the same thing happend. I think this code go against this rule and there not a way to prevent this bad practice. 

I saw this pattern multiple times in a lot of code bases, I try to enable this rule but just a few cases was detected, and all the other that use the name "render" in they function will take like actual valid code.

I'd be happy to contribute, if we defined which will be the best option to follow

**Comment by DaniAcu** [comment]
I would like to re-open this issue. I found many codebases written with that rule activeted that is writen in that way.

```jsx
function Uploader ({ readOnly }) {
  function readOnlyMessage() {
    // checking some state and render jsx or null
  }

  function readOnlyFileList() {
    // render jsx
  }

  function renderPreloadedItems() {
   return scopedState.map(something => {
      return <Wrapper {...aLotOfPropsConfig}>{readOnlyFileList()}</Wrapper>
   })
  }
  
  function renderUploader() {
   return (
     <UiComponent {...propsAssignedBasedOfState}>
     {renderPreloadedItems()}
     {scopedState.map(moreJSXHere)}
     </UiComponent>
  )
  }

  return readOnly ? (
      <>
        {readOnlyMessage()}
        {renderPreloadedItems()}
      </>
    ) : (
      renderUploader()
    );
}
```

If this functions are written with title case like `RenderUploader` it will trigger a error, but for now it will not and it allowed to the unexperience dev to bypass the rule. I try to disable the allowedProp to see if in that case it will be throwing some error but still taking this as valid code. I also create a unit test in this repo and the same thing happend. I think this code go against this rule and there not a way to prevent this bad practice. 

I saw this pattern multiple times in a lot of code bases, I try to enable this rule but just a few cases was detected, and all the other that use the name "render" in they function will take like actual valid code.

I'd be happy to contribute, if we defined which will be the best option to follow

**Comment by ljharb** [comment]
The issue there is not calling functions - it's defining those functions inside the component.

**Comment by FezVrasta** [comment]
The same would apply if the functions were defined outside the component definition.

**Comment by ljharb** [comment]
Can you elaborate on what issue that would cause? It's just like any code - it's fine, and often ideal, to extract code into a separate function for reuse, or readability.

**Comment by FezVrasta** [comment]
I think I explained it above, it bails the rendering of that function out from the React reconciler.

**Comment by ljharb** [comment]
no it doesn't? The reconciler has no visibility into how you organize your code, because of how JavaScript works.

**Comment by FezVrasta** [comment]
`<MyFunction />` is translated into `React.createElement(MyFunction, ...)`, `muFunction()` is not touched, so it isn't considered a component and the reconciler can't do anything with it.

**Comment by ljharb** [comment]
I'm very confused by your comment. You can create elements at any time; the reconciler only cares about attempting to *render* those elements. The source of the element is irrelevant.

**Comment by DaniAcu** [comment]
@ljharb I think what @FezVrasta wants to say is that the calling of a function in the render like that

```js
function Component () {
  return <Wrapper>{renderSomething()</Wrapper>

}

```

The render someting will not work as a component, because the reconciler they will re-render everytime that the parent component is re-render. Despite the `renderSometing` has a state to manage or not.

I mean if you go to the code example in the docs of this rule and change everything with renderSometing, the bug that you show will still happening, which I think is a way to bypass this rule.ñ becuase this rule will not thow.

For another hand, i think that you agree that the creation of functions inside components in the way i show is bad, maybe is not for this rule, but we could create other to track that? Or there anything already created that i could use it to prevent it?

**Comment by ljharb** [comment]
@DaniAcu sure. but the thing that `renderSomething` RETURNS is what matters, and absolutely might work as a component. It's a JavaScript function - if you inline the contents of `renderSomething` it behaves absolutely identically.

**Comment by FezVrasta** [comment]
I'll try one last time. Inexperienced developers don't realize that they are missing out in possible performance optimizations by doing `renderSomething()` rather than `<Something />`. The requested rule would help them realize this.

**Comment by ljharb** [comment]
They aren't missing out on *any* possible optimizations, though, that's what I'm not understanding. The reason to use components instead of render functions is clarity - a subjective thing - and not because it has any impact on react's renderer.

**Comment by FezVrasta** [comment]
I don't know how else to explain it, I'm sorry.

**Comment by ljharb** [comment]
If you can provide a runnable example that shows an optimization opportunity that this proposed rule would help with, then I'd love to see it - otherwise, my understanding of React leads me to believe that you're incorrect.

**Comment by dacunaDesp** [comment]
>  but the thing that renderSomething RETURNS is what matters, and absolutely might work as a component. 

I think there a incongruense in the rule. 

This will be warning
```jsx
function ParentComponent() {
   function UnsableComponent() {
     return <div />;
   }

    return (
     <div>
       {UnsableComponent()}
     </div>
    );
}
 ```
 
 but this not?
 
 ```jsx
function ParentComponent() {
   function unsableComponent() {
     return <div />;
   }

    return (
     <div>
       {unsableComponent()}
     </div>
    );
}
 ```
 
 As far as I know of React both have the same impact, why we warning about one pattern and not another?

**Comment by ljharb** [comment]
@dacunaDesp yes, that's right. They are not the same - the second example is a perfectly reasonable way to reuse an encapsulate code. The first *would* be, except that the component isn't being rendered, it's being invoked - and *that*, alone, is what the current rule warns about.

**Comment by AriPerkkio** [comment]
> As far as I know of React both have the same impact, why we warning about one pattern and not another?

Because React does not allow user to render custom (non-host) components if they are lower case. You cannot render `<unsableComponent />` but you can render `<UnsableComponent />`. And when the component is nested, it's re-created on each re-render and the function reference changes. Reconciler sees it as new component and discards the component tree completely. Note that the rule warns about component declaration, not how it's used.

When you render `{unsableComponent()}` or the upper case one, all React sees is that there is `{ type: "div" }` - at least in React 16, I haven't debugged the latest versions.

Whether it should be allowed to create render functions inside React component and call these functions, that is a stylistic issue. It does not affect how reconciler sees it. Such stylistic rule would be something I would enable in my codebases.

**Comment by ljharb** [comment]
I certainly can see the value of a rule that forces moving such a render function to module level, but I don't think it's a rule that could be useful and reliable in practice for the stylistic aspect.

**Comment by DaniAcu** [comment]
Oh!! This rule is just to prevent us `<UnsableComponent />` as a component when it really just work as `{render()}`, and the reconciler just tale that as part of the parent component. I understand correctly?

So this just prevent an unexpected result of the use `<UnsableComponent />`.

Thanks for the patient of explain that.

And for stylistic things like preferences of one or another to prevent the over use of this pattern, as I show before. We have some repo for that or we could create another rule. Or you think that it's better not have force stylish alignment in proyects?

**Comment by DaniAcu** [comment]
@ljharb why do you think the rule will not be useful or reliable?

I just see a lot of bad use cases that use this pattern, despite this doesn't have performance changes, with bad abstraction just because it easy use the state here and not create a extra component. I would like have a rule that force the stylish with a good message like "move that outside of the render function".

I think it's possible, but maybe im just simplify the complexity of the task.

**Comment by ljharb** [comment]
For one, because a component can return null, a string, an array, a number, etc, so *any function* returning that would be warned on by this rule.

Most eslint plugins only check one file at a time, so this wouldn't work *at all* for functions defined in other files.

What about hooks, like custom hooks? What about the perfectly valid pattern of HOCs - higher order components? There's lots of ways it makes sense to call a function and render the result, and it would be _impossible_ for a linter to know when those ways applied.

**Comment by DaniAcu** [comment]
Thanks @ljharb I was thinking in some of that cases. HOC and hooks that use render methods is right.

But what if we check that a function that return a jsx in some brach, can't have another function that returns a jsx declared inside.

I think that it's possible. But what do you think?

**Comment by ljharb** [comment]
Since it wouldn't work for functions declared elsewhere, I don't think it would be particularly valuabl

**Comment by DaniAcu** [comment]
Wait but if this rule work in each file, we garantee that every function it will not have a render method nested in another render method. At least every function written in the project where this rule is enabled.

**Comment by ljharb** [comment]
I don’t think i agree with your last statement, sometimes render functions are across files.

**Comment by DaniAcu** [comment]
The pattern that we are talking about is render function declared inside another render method.
Right?

```jsx

function Component ({ is }) {
  function renderA() { /* jsx and managed scope state */ }

  function renderB() { /* jsx */ }

  if (is === "A") return renderA();
  if (is === "B") return renderB();

  return null;

}

```

As you said it will be helpful move this render methods to module level.

"this render methods" are declared inside of function render component.

So as this is declared inside of component could be detected in one file, because is declared inside of another so in the same module.

If this is use in module level, i think we agree before it is okay.

The only corner case I could think is using HOC inside render method. 

```jsx

import { createRenderA } from "renderA";

import { createRenderB } from "renderB";

function Component ({ is }) {
  const renderA = createRenderA();
  const renderB = createRenderB();

  if (is === "A") return renderA();
  if (is === "B") return renderB();

  return null;
}

```

But this is another case and also a pattern where HOC is not commonly used and using a HOC doesn't allow to use scoped state or setState, which could be confused to follow (this is just a stylish preference).

In the previous case they declare the function inside to manipulate the state incide of scope. To prevent this style we could detect where render is declared inside and force to move to top level. 

Every code I saw just use the first pattern that I present.

**Comment by ljharb** [comment]
I very commonly use an HOC pattern where the factory is stored in a separate module.

As I said earlier, a rule that forces these kinds of functions to be created outside a component would be great - that's not stylistic, that's just about correctness and performance. A rule that tries to prevent any non-component functions from producing jsx seems to be quite extreme and more harmful than helpful.

**Comment by damdeez** [comment]
Related: https://react.dev/reference/rules/react-calls-components-and-hooks#never-call-component-functions-directly
